analyzers/src/SonarAnalyzer.Common/Rules/ExpressionComplexityBase.cs

```diff
@@ -30,39 +30,57 @@ public abstract class ExpressionComplexityBase<TExpression> : ParameterLoadingDi
+            context.RegisterSyntaxNodeActionInNonGenerated(Language.GeneratedCodeRecognizer, c =>
                 {
-                    var root = c.Tree.GetRoot();
-                    var rootExpressions = NoncompoundSubexpressions(root);
-                    var compoundExpressionsDescendants = root.DescendantNodes().Where(IsCompoundExpression).SelectMany(NoncompoundSubexpressions);
-
-                    foreach (var expression in rootExpressions.Concat(compoundExpressionsDescendants))
+                    if (NodeKind(c.Node.Parent) is TSyntaxKind parentKind && (ComplexityIncreasingKinds.Contains(parentKind) || TransparentKinds.Contains(parentKind)))
+                    {
+                        // The parent of the expression is itself complexity increasing (e.g. &&) or transparent (e.g. parenthesis).
+                        // We are only interested in the expression roots so we only report once per expression tree. Therefore we ignore any inner children, e.g.:
+                        // a && b && c -> (Left: (Left: a, Right: b), Right: c) // We are only interested in the outer expression (the one with Right: c)
+                        return;
+                    }
+                    var complexity = CalculateComplexity(c.Node);
+                    if (complexity > Maximum)
                     {
-                        var complexity = expression.DescendantNodesAndSelf(x => !IsCompoundExpression(x)).Count(IsComplexityIncreasingKind);
-                        if (complexity > Maximum)
-                        {
-                            c.ReportIssue(Diagnostic.Create(rule, expression.GetLocation(), Maximum, complexity));
-                        }
+                        c.ReportIssue(Diagnostic.Create(rule, c.Node.GetLocation(), Maximum, complexity));
                     }
-                });
+                }, ComplexityIncreasingKinds.Concat(TransparentKinds).ToArray());
+
+        private static TSyntaxKind? NodeKind(SyntaxNode node) =>
+            node == null
+                ? null
+                : Enum.ToObject(typeof(TSyntaxKind), node.RawKind) as TSyntaxKind?;
 
-        private IEnumerable<SyntaxNode> NoncompoundSubexpressions(SyntaxNode node) =>
-            node.DescendantNodes(x => !IsExpressionOrPatternRoot(x) || x == node).Where(x => IsExpressionOrPatternRoot(x) && !IsCompoundExpression(x));
+        private int CalculateComplexity(SyntaxNode node)
+        {
+            return CalculateComplexityRec(node, 0);
+
+            int CalculateComplexityRec(SyntaxNode node, int currentComplexity)
+            {
+                if (NodeKind(node) is TSyntaxKind kind && ComplexityIncreasingKinds.Contains(kind))
+                {
+                    currentComplexity++;
+                }
+                foreach (var child in ExpressionChildren(node))
+                {
+                    currentComplexity = CalculateComplexityRec(child, currentComplexity);
```

Comment:
Educational (and maybe dumb) question, why this is not:
`currentComplexity = Math.Max(currentComplexity, CalculateComplexityRec(child, currentComplexity));`

Reply:
currentComplexity either stays the same or increases. There is no need for "Max". If you think it is confusing to pass the "currentCompelxity" around, we could change it like this:
```cs
        private int CalculateComplexity(SyntaxNode node)
        {
            int currentComplexity = 0;
            CalculateComplexityRec(node);
            return currentComplexity;

            int CalculateComplexityRec(SyntaxNode node)
            {
                if (NodeKind(node) is TSyntaxKind kind && ComplexityIncreasingKinds.Contains(kind))
                {
                    currentComplexity++;
                }
                foreach (var child in ExpressionChildren(node))
                {
                    CalculateComplexityRec(child);
                }
            }
        }
```

Do you like this version better?

Reply:
I think this version is easier to grasp. Thanks for the explanation